tools/ci-build/runtime-versioner/src/audit.rs

```diff
@@ -0,0 +1,260 @@
+            Some(1) => Ok(true),
+            code => bail!("unknown git diff result: {code:?}"),
+        }
+    }
+}
+
+/// Loads version information from crates.io and attaches it to the passed in runtime crates.
+fn augment_runtime_crates(
+    previous_crates: BTreeMap<String, DiscoveredCrate>,
+    next_crates: BTreeMap<String, DiscoveredCrate>,
+    fake_crates_io_index: Option<Utf8PathBuf>,
+) -> Result<Vec<RuntimeCrate>> {
+    let index = fake_crates_io_index
+        .map(CratesIndex::fake)
+        .map(Ok)
+        .unwrap_or_else(CratesIndex::real)?;
+    let all_keys: BTreeSet<_> = previous_crates.keys().chain(next_crates.keys()).collect();
+    let mut result = Vec::new();
+    for key in all_keys {
+        let previous_crate = previous_crates.get(key);
+        if let Some(next_crate) = next_crates.get(key) {
+            result.push(RuntimeCrate {
+                published_versions: index.published_versions(&next_crate.name)?,
+                name: next_crate.name.clone(),
+                previous_release_version: previous_crate.map(|c| c.version.clone()),
+                next_release_version: next_crate.version.clone(),
+                path: next_crate.path.clone(),
+            });
+        } else {
+            tracing::warn!("runtime crate '{key}' was removed and will not be published");
+        }
+    }
+    Ok(result)
+}
+
+struct DiscoveredCrate {
+    name: String,
+    version: String,
+    path: Utf8PathBuf,
+}
+
+/// Discovers runtime crates that are independently versioned.
+/// For now, that just means the ones that don't have the special version number `0.0.0-smithy-rs-head`.
+/// In the future, this can be simplified to just return all the runtime crates.
+fn discover_runtime_crates(repo_root: &Utf8Path) -> Result<BTreeMap<String, DiscoveredCrate>> {
+    const ROOT_PATHS: &[&str] = &["rust-runtime", "aws/rust-runtime"];
+    let mut result = BTreeMap::new();
+    for &root in ROOT_PATHS {
+        let root = repo_root.join(root);
+        for entry in fs::read_dir(&root)
+            .context(root)
+            .context("failed to read dir")?
+        {
+            let entry = entry.context("failed to read dir entry")?;
+            if !entry.path().is_dir() {
+                continue;
+            }
+            let manifest_path = entry.path().join("Cargo.toml");
+            if !manifest_path.exists() {
+                continue;
+            }
+            let manifest: toml::Value =
+                toml::from_slice(&fs::read(&manifest_path).context("failed to read manifest")?)
+                    .context("failed to parse manifest")?;
+            let publish = manifest["package"]
+                .get("publish")
+                .and_then(|p| p.as_bool())
+                .unwrap_or(true);
+            let version = manifest["package"]["version"]
+                .as_str()
+                .expect("version is a string");
+            if publish && version != "0.0.0-smithy-rs-head" {
```

Comment:
I'm _slightly_ concerned about this check going in this direction. Is there anything that mandates those versions? Just want to make sure we don't have a crate with `0.0.0-smithy-rs` or something slightly different by accident.

Side note...what if we set this to `0.0.0-smithy-rs-stable`, `0.0.0-smithy-rs-unstable`....

Reply:
> I'm slightly concerned about this check going in this direction. Is there anything that mandates those versions? Just want to make sure we don't have a crate with 0.0.0-smithy-rs or something slightly different by accident.

It's a fair concern, and if we were going to keep this situation forever, we might want to think of something to avoid it. But I think we should push for getting rid of the stable/unstable crate versions soon after independent runtime crate versions have been proven.

> Side note...what if we set this to 0.0.0-smithy-rs-stable, 0.0.0-smithy-rs-unstable....

It's an interesting idea. Maybe we should have done it that way to begin with a couple months back. I'm not sure it's worth the investment to refactor now though.

Reply:
reasonable — maybe we should have a check that it's either 0.0.0-... or a real semver version